src/sound/snd_opl_esfm.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * 86Box    A hypervisor and IBM PC system emulator that specializes in
+ *          running old operating systems and software designed for IBM
+ *          PC systems and compatibles from 1981 through fairly recent
+ *          system designs based on the PCI bus.
+ *
+ *          This file is part of the 86Box distribution.
+ *
+ *          ESFMu ESFM emulator.
+ * 
+ * 
+ * Authors: Fred N. van Kempen, <decwiz@yahoo.com>
+ *          Miran Grca, <mgrca8@gmail.com>
+ *          Alexey Khokholov (Nuke.YKT)
+ *          Cacodemon345
+ *
+ *          Copyright 2017-2020 Fred N. van Kempen.
+ *          Copyright 2016-2020 Miran Grca.
+ *          Copyright 2013-2018 Alexey Khokholov (Nuke.YKT)
+ *          Copyright 2024 Cacodemon345
+ */
+
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "esfmu/esfm.h"
+
+#define HAVE_STDARG_H
+#define NO_SOFTFLOAT_INCLUDE
+#include <86box/86box.h>
+#include <86box/sound.h>
+#include <86box/device.h>
+#include <86box/timer.h>
+#include <86box/snd_opl.h>
+
+#define RSM_FRAC    10
+
+enum {
+    FLAG_CYCLES = 0x02,
+    FLAG_OPL3   = 0x01
+};
+
+typedef struct {
+    esfm_chip opl;
+    int8_t  flags;
+    int8_t  pad;
+
+    uint16_t port;
+    uint8_t  status;
+    uint8_t  timer_ctrl;
+    uint16_t timer_count[2];
+    uint16_t timer_cur_count[2];
+
+    // OPL3L
+    int32_t rateratio;
+    int32_t samplecnt;
+    int32_t oldsamples[2];
+    int32_t samples[2];
+
+    int     pos;
+    int32_t buffer[SOUNDBUFLEN * 2];
+} esfm_drv_t;
+
+void
+esfm_drv_generate_resampled(esfm_drv_t *dev, int32_t *bufp)
+{
+    while (dev->samplecnt >= dev->rateratio) {
+        int16_t samples[2] = { 0, 0 };
+        dev->oldsamples[0] = dev->samples[0];
+        dev->oldsamples[1] = dev->samples[1];
+        ESFM_generate(&dev->opl, samples);
+        dev->samples[0] = samples[0];
+        dev->samples[1] = samples[1];
+        dev->samplecnt -= dev->rateratio;
+    }
+
+    bufp[0] = (int32_t) ((dev->oldsamples[0] * (dev->rateratio - dev->samplecnt)
+                          + dev->samples[0] * dev->samplecnt)
+                         / dev->rateratio);
+    bufp[1] = (int32_t) ((dev->oldsamples[1] * (dev->rateratio - dev->samplecnt)
+                          + dev->samples[1] * dev->samplecnt)
+                         / dev->rateratio);
```

Comment:
Unlike Yamaha OPL3 chips, no ESFM chips ever did linear resampling of the output - they all output natively at ~49716 Hz. Please don't linearly resample the output.

Reply:
The emulator can't handle sound sources with sample rates different than 48000 Hz unfortunately, so I had no other option.

Reply:
Hmm, 48 KHz from 49.7 KHz isn't that bad for linear resampling, and ESFM generates much less feedback whine compared to Yamaha FM, so I guess I'll let it pass. Oh well.

Reply:
That can, and should, be rectified - we need a separate sound source for OPL so we can do the correct sample rate.

Reply:
This is now done. We now have a 49716 Hz source so ESFM can now be done without resampling.